sdk_contrib/aws_v1_support/src/main/java/io/opentelemetry/sdk/contrib/trace/aws/resource/Ec2ResourcePopulator.java

```diff
@@ -0,0 +1,207 @@
+
+  private static final int TIMEOUT_MILLIS = 2000;
+
+  private static final JsonFactory JSON_FACTORY = new JsonFactory();
+
+  private static final String DEFAULT_IMDS_ENDPOINT = "169.254.169.254";
+
+  private final URL identityDocumentUrl;
+  private final URL hostnameUrl;
+  private final URL tokenUrl;
+
+  Ec2ResourcePopulator() {
+    // This is only for testing e.g., with a mock IMDS server and never in production so we just
+    // read from a system property. This is similar to the AWS SDK.
+    this(System.getProperty("otel.aws.imds.endpointOverride", DEFAULT_IMDS_ENDPOINT));
+  }
+
+  @VisibleForTesting
+  Ec2ResourcePopulator(String endpoint) {
+    String urlBase = "http://" + endpoint;
+    try {
+      this.identityDocumentUrl = new URL(urlBase + "/latest/dynamic/instance-identity/document");
+      this.hostnameUrl = new URL(urlBase + "/latest/meta-data/hostname");
+      this.tokenUrl = new URL(urlBase + "/latest/api/token");
+    } catch (MalformedURLException e) {
+      // Can only happen when overriding the endpoint in testing so just throw.
+      throw new IllegalArgumentException("Illegal endpoint: " + endpoint, e);
+    }
+  }
+
+  @Override
+  public void populate(Map<String, AttributeValue> resourceAttributes) {
+    String token = fetchToken();
+
+    // If token is empty, either IMDSv2 isn't enabled or an unexpected failure happened. We can
+    // still get data if IMDSv1 is enabled.
+    String identity = fetchIdentity(token);
+    if (identity.isEmpty()) {
+      // If no identity document, assume we are not actually running on EC2.
+      return;
+    }
+
+    String hostname = fetchHostname(token);
+
+    try (JsonParser parser = JSON_FACTORY.createParser(identity)) {
+      parser.nextToken();
+
+      if (!parser.isExpectedStartObjectToken()) {
+        throw new IOException("Invalid JSON:" + identity);
+      }
+
+      while (parser.nextToken() != JsonToken.END_OBJECT) {
+        String value = parser.nextTextValue();
+        switch (parser.getCurrentName()) {
+          case "instanceId":
+            resourceAttributes.put(ResourceConstants.HOST_ID, stringAttributeValue(value));
+            break;
+          case "availabilityZone":
+            resourceAttributes.put(ResourceConstants.CLOUD_ZONE, stringAttributeValue(value));
+            break;
+          case "instanceType":
+            resourceAttributes.put(ResourceConstants.HOST_TYPE, stringAttributeValue(value));
+            break;
+          case "imageId":
+            resourceAttributes.put(ResourceConstants.HOST_IMAGE_ID, stringAttributeValue(value));
+            break;
+          case "accountId":
+            resourceAttributes.put(ResourceConstants.CLOUD_ACCOUNT, stringAttributeValue(value));
+            break;
+          case "region":
+            resourceAttributes.put(ResourceConstants.CLOUD_REGION, stringAttributeValue(value));
+            break;
+          default:
```

Comment:
It is missing the case for "privateIp" to populate `HOST_NAME`

Reply:
Yeah asked in my PR description about that - looking at the documentation for HOST_NAME, it didn't seem like it's meant for an IP address so left it out for now. But is it better to set that to the IP address?